metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala

```diff
@@ -102,33 +120,103 @@ class DebugProvider(
   )
 
   def start(
-      parameters: b.DebugSessionParams,
-      cancelPromise: Promise[Unit],
+      parameters: b.DebugSessionParams
   )(implicit ec: ExecutionContext): Future[DebugServer] = {
+    val cancelPromise = Promise[Unit]()
     for {
       sessionName <- Future.fromTry(parseSessionName(parameters))
       jvmOptionsTranslatedParams = translateJvmParams(parameters)
       buildServer <- buildServerConnect(parameters)
         .fold[Future[BuildServerConnection]](BuildServerUnavailableError)(
           Future.successful
         )
-      debugServer <- start(
+      isJvm = parameters
+        .getTargets()
+        .asScala
+        .flatMap(buildTargets.scalaTarget)
+        .forall(
+          _.scalaInfo.getPlatform == b.ScalaPlatform.JVM
+        )
+      debugServer <-
+        if (isJvm)
+          statusBar.trackSlowFuture(
+            "Starting debug server",
+            start(
+              sessionName,
+              jvmOptionsTranslatedParams,
+              buildServer,
+              cancelPromise,
+            ),
+            () => cancelPromise.trySuccess(()),
+          )
+        else
+          runLocally(
+            sessionName,
+            jvmOptionsTranslatedParams,
+            buildServer,
+            cancelPromise,
+          )
+    } yield debugServer
+  }
+
+  private def runLocally(
+      sessionName: String,
+      parameters: b.DebugSessionParams,
+      buildServer: BuildServerConnection,
+      cancelPromise: Promise[Unit],
+  )(implicit ec: ExecutionContext): Future[DebugServer] = {
+    val proxyServer = new ServerSocket(0, 50, localAddress)
+    val host = InetAddresses.toUriString(proxyServer.getInetAddress)
+    val port = proxyServer.getLocalPort
+    proxyServer.setSoTimeout(10 * 1000)
+    val uri = URI.create(s"tcp://$host:$port")
+
+    val awaitClient = () => Future(proxyServer.accept())
+
+    DebugRunner
+      .open(
         sessionName,
-        jvmOptionsTranslatedParams,
-        buildServer,
+        awaitClient,
+        stacktraceAnalyzer,
+        () => {
+          val runParams = new b.RunParams(parameters.getTargets().asScala.head)
+          if (
+            parameters
+              .getDataKind() == b.DebugSessionParamsDataKind.SCALA_MAIN_CLASS
+          ) {
+            runParams.setDataKind(parameters.getDataKind())
+            runParams.setData(parameters.getData())
+          }
+          buildServer.buildTargetRun(runParams, cancelPromise)
+        },
         cancelPromise,
       )
-    } yield debugServer
+      .flatMap { runner =>
+        currentRunner.set(runner)
```

Comment:
what will happen if I run 2 different programs? Let assume that first one is long running computation and in the meantime I'll run another one.  Will it mess logging or is it negligible corner case? I checked vscode now and I can debug two different programs 🤔

Reply:
> negligible corner case

I guess number of people trying to run on js and native platforms is already negligible 😅

Reply:
I don't have a way currently to differentiate between the two programs, so this will not be possible. We can make it explicit.

There was a discussion in BSP about it, but I can't find it currently. We could  identify each log message by the origin id maybe, let me check it.

Reply:
> We could identify each log message by the origin id maybe, let me check it.

just have in mind that this is probably minor issue and might not be worth spending too much time on it ;)

Reply:
I added an atomic boolean that should take care of it. It will be set to false only if run ends or the user disconnects, but that should work fine.

I also realized that we don't show errors proeprly if code lense failed to run, so I added that in the PR here: https://github.com/scalameta/metals-vscode/pull/1449

Reply:
atomic flag might not be the best looking solution, but it's a working one. Let's get this thing merged